### PR TITLE
feat: implement state storage

### PR DESCRIPTION
**Pull request by utkuozdemir**
On a best-effort basis, store the state on the disk periodically and on shutdown & restore it from the disk on startup.

Additionally, bump Go version, deps & rekres.

Closes siderolabs/discovery-service#54.

### CODE REVIEW

**Comment by nberlee** [comment]
Consider removing
```
Moreover, the discovery service has no peristence. Data is stored in memory only [...]
```
from `internal/landing/html/index.html`

**Comment by utkuozdemir** [comment]
> Consider removing
> 
> ```
> Moreover, the discovery service has no peristence. Data is stored in memory only [...]
> ```
> 
> from `internal/landing/html/index.html`

Good one, thank you!

File: cmd/snapshot-decoder/main.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+	snapshot := &storagepb.StateSnapshot{}
+
+	if err = snapshot.UnmarshalVT(data); err != nil {
+		return fmt.Errorf("failed to unmarshal snapshot: %w", err)
+	}
+
+	encoder := json.NewEncoder(os.Stdout)
+
+	encoder.SetIndent("", "  ")
```
Comment (Unix4ever) [inline cmd/snapshot-decoder/main.go:47]: Is it really necessary?

File: internal/state/storage/storage.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+			if err := storage.Save(); err != nil {
+				storage.logger.Error("failed to save state", zap.Error(err))
+			}
+		}
+	}
+}
+
+// Save saves all clusters' states into the persistent storage.
+func (storage *Storage) Save() (err error) {
+	if storage.inProgress {
```
Comment (Unix4ever) [inline internal/state/storage/storage.go:145]: I'm confused about this flag: it looks like this Save runs in the same goroutine and it blocks until it's done.
Why do we need this flag then?

On other hand if it can run in different goroutines then we should have thread safety here.

File: Dockerfile [hunk 3]
```diff
@@ -135,3 +190,11 @@ COPY --from=image-ca-certificates / /
 LABEL org.opencontainers.image.source https://github.com/siderolabs/discovery-service
 ENTRYPOINT ["/discovery-service"]
 
+FROM scratch AS image-snapshot-decoder
```
Comment (smira) [inline Dockerfile:193]: we need to disable this image in Kres I guess

File: internal/state/snapshot.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+func (state *State) Snapshot() *storagepb.StateSnapshot {
+	var clusterSnapshots []*storagepb.ClusterSnapshot
+
+	state.clusters.Range(func(_ string, cluster *Cluster) bool {
+		clusterSnapshots = append(clusterSnapshots, snapshotCluster(cluster))
+
+		return true
+	})
+
+	return &storagepb.StateSnapshot{
```
Comment (smira) [inline internal/state/snapshot.go:28]: can we do a benchmark of how many allocations happen on this path -- from the state snapshot to the final on-disk representation. 

I have some fears about protobuf

File: internal/state/snapshot.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	}
+}
+
+func snapshotAffiliate(affiliate *Affiliate) *storagepb.AffiliateSnapshot {
+	return &storagepb.AffiliateSnapshot{
+		Id:         affiliate.id,
+		Expiration: timestamppb.New(affiliate.expiration),
+		Data:       affiliate.data,
+		Endpoints:  xslices.Map(affiliate.endpoints, snapshotEndpoint),
+		Changed:    affiliate.changed,
```
Comment (smira) [inline internal/state/snapshot.go:60]: this is a temporary flag used internally, I don't think we should snapshot it

File: internal/state/snapshot.go (hunk 4 above)
Comment (smira) [reply internal/state/snapshot.go:28]: plus it almost feels like we could do better job if we write cluster-by-cluster without trying to put all clusters at the same time to memory.

so there's no `StateSnapshot`, but a list of protobufs for each `Cluster`

File: api/storage/storage.proto [hunk 6]
```diff
@@ -0,0 +1,34 @@
+syntax = "proto3";
+
+package sidero.discovery.storage;
+option go_package = "github.com/siderolabs/discovery-service/api/storagepb";
+
+import "google/protobuf/timestamp.proto";
+
+// StateSnapshot is the snapshot of the discovery service state.
```
Comment (smira) [inline api/storage/storage.proto:8]: I'm torn on that, but feels like it _might_ make sense to move this down to https://github.com/siderolabs/discovery-api as a storage/snapshot proto.

File: api/storage/storage.proto [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+//
+// We avoid marshalling/unmarshalling to this type directly, as it causes an allocation of all of the clusters at once.
```
Comment (smira) [inline api/storage/storage.proto:10]: it feels like this comment doesn't belong here? it might be better in the Marshal/Unmarshal methods, as it doesn't change the proto definition

File: cmd/snapshot-decoder/main.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+		return fmt.Errorf("failed to read snapshot: %w", err)
+	}
+
+	snapshot := &storagepb.StateSnapshot{}
+
+	if err = snapshot.UnmarshalVT(data); err != nil {
+		return fmt.Errorf("failed to unmarshal snapshot: %w", err)
+	}
+
+	encoder := json.NewEncoder(os.Stdout)
```
Comment (smira) [inline cmd/snapshot-decoder/main.go:45]: we might consider https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson

File: Dockerfile (hunk 3 above)
Comment (smira) [reply Dockerfile:193]: bump this!

File: internal/landing/html/index.html [hunk 9]
```diff
@@ -34,12 +34,6 @@ <h2>Details</h2>
             Each peer receives information back about other affiliates from the discovery service, decrypts it and uses it to drive KubeSpan and cluster discovery.
         </p>
 
-        <p>
-            Moreover, the discovery service has no peristence.
```
Comment (smira) [inline internal/landing/html/index.html:38]: I think we dropped too much here. The first sentence doesn't apply, but the second might be rephrased with `Data is stored in memory with TTL, and periodic snapshots to disk are enabled.`

File: internal/state/snapshot.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+			snapshot.Affiliates[i] = &storagepb.AffiliateSnapshot{}
+		}
+
+		snapshot.Affiliates[i].Id = affiliate.id
+
+		if snapshot.Affiliates[i].Expiration == nil {
+			snapshot.Affiliates[i].Expiration = &timestamppb.Timestamp{}
+		}
+
+		snapshot.Affiliates[i].Expiration.Seconds = affiliate.expiration.Unix()
```
Comment (smira) [inline internal/state/snapshot.go:81]: there should be a helper in `timestamppb` package to do conversion between protobuf and time.Duration

File: internal/state/storage/protobuf.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+	clustersFieldNum  = 1
+	clustersFieldType = protowire.BytesType
+)
+
+// encodeClusterSnapshot encodes a ClusterSnapshot into the given buffer, resizing it as needed.
+//
+// It returns the buffer with the encoded ClusterSnapshot and an error if the encoding fails.
+func encodeClusterSnapshot(buffer []byte, snapshot *storagepb.ClusterSnapshot) ([]byte, error) {
+	buffer = protowire.AppendTag(buffer, clustersFieldNum, clustersFieldType)
+	buffer = protowire.AppendVarint(buffer, uint64(snapshot.SizeVT()))
```
Comment (smira) [inline internal/state/storage/protobuf.go:27]: looks like we call `SizeVT()` twice, we could use it just once

File: internal/state/storage/storage.go [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+	buffer := make([]byte, bufferSize)
+
+	// unmarshal the clusters in a streaming manner and import them into the state
+	if err := storage.state.ImportClusterSnapshots(func() (*storagepb.ClusterSnapshot, bool, error) {
+		for {
+			numRead, err := reader.Read(buffer[index:])
+			size += numRead
+
+			if err != nil {
+				if err == io.EOF { //nolint:errorlint
```
Comment (smira) [inline internal/state/storage/storage.go:250]: there is a slight problem here - as we  can come here both when we're at the middle of the cluster (when reading more) and when we are in between reads, one of the EOFs is fine and another is unexpected.

I would instead rework this code to simplify it:


* use `bufio.Reader` to avoid hitting actual reader with many reads
* read enough bytes to read the protobuf type/tag/size, decode actual size and more bytes to read 
* read exactly required amount of bytes, and pass it to decoder

Basically split reading header which has fixed max size and other (unknown) size part into two separate reads, but amortized via `bufio.Reader`.

Use `io.ReadFull` to make sure we can have full buffer, no matter what Reader returns.

File: internal/state/storage/storage.go [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+	return SnapshotStats{
+		Size:          size,
+		NumClusters:   numClusters,
+		NumAffiliates: numAffiliates,
+		NumEndpoints:  numEndpoints,
+	}, nil
+}
+
+func getTempFile(dst string) (*os.File, error) {
+	tmpFile, err := os.OpenFile(dst+".tmp", os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o666)
```
Comment (smira) [inline internal/state/storage/storage.go:348]: if we use `O_EXCL`, we can easily break things if there's a `.tmp` file already in the filesystem (e.g. if crashed).

I would rather use `O_TRUNC`.

File: Dockerfile (hunk 3 above)
Comment (utkuozdemir) [reply Dockerfile]: done

Comment (frezbo) [reply Dockerfile]: also needs a rekres now, since we changed the builders

Comment (utkuozdemir) [reply Dockerfile]: done

File: internal/state/storage/storage.go [hunk 14]
```diff
@@ -0,0 +1,414 @@
...
+	size := 0
+	numClusters := 0
+	numAffiliates := 0
+	numEndpoints := 0
+	buffer := make([]byte, 0, maxHeaderSize)
+	reader = bufio.NewReader(reader)
+
+	// unmarshal the clusters in a streaming manner and import them into the state
+	if err := storage.state.ImportClusterSnapshots(func() (*storagepb.ClusterSnapshot, bool, error) {
+		var initialNumRead, additionalNumRead int
```
Comment (smira) [inline internal/state/storage/storage.go:248]: I guess this can be simplified more, probably we can do a quick hack session?

File: internal/state/storage/storage.go [hunk 15]
```diff
@@ -0,0 +1,414 @@
...
+		}
+
+		// trim the buffer to the actual read size
+		buffer = buffer[:readStart+initialNumRead]
+
+		if len(buffer) == 0 { // processed all data, return
+			return nil, false, nil
+		}
+
+		headerSize, clusterSize, err := decodeClusterSnapshotHeader(buffer)
```
Comment (smira) [inline internal/state/storage/storage.go:266]: we need to have a maximum supported `clusterSize`

File: internal/state/storage/storage.go (hunk 14 above)
Comment (utkuozdemir) [reply internal/state/storage/storage.go]: Updated one more time after our discussion.

File: internal/state/storage/storage.go (hunk 15 above)
Comment (utkuozdemir) [reply internal/state/storage/storage.go]: Done, also added a test where I create a cluster with the maximum possible marshaled size within the discovery service limits and test importing it when it's just at the max size, and when it is a bit larger than that.

I also use the computed max size from that test as the max allowed value.

**Review by smira — APPROVED** [review]
:cool: :rocket:

File: internal/state/storage/storage.go [hunk 16]
```diff
@@ -0,0 +1,391 @@
...
+		NumAffiliates: numAffiliates,
+		NumEndpoints:  numEndpoints,
+	}, nil
+}
+
+// Export exports all clusters' states into the given writer.
+//
+// When exporting, we avoid marshaling to the storagepb.StateSnapshot type directly, as it causes an allocation of all the cluster snapshots at once.
+// Instead, we process clusters in a streaming manner, marshaling them one by one and exporting them into the writer.
+func (storage *Storage) Export(writer io.Writer) (SnapshotStats, error) {
```
[the remark this answers is not in the document]
Comment (utkuozdemir) [reply internal/state/storage/storage.go:299]: makes sense, done
